virtio-devices/src/balloon.rs

```diff
@@ -165,6 +211,55 @@ impl BalloonEpollHandler {
         Self::advise_memory_range(memory, range_base, range_len, libc::MADV_DONTNEED)
     }
 
+    fn release_memory_range_4k(
+        pbp: &mut Option<PartiallyBalloonedPage>,
+        memory: &GuestMemoryMmap,
+        pfn: u32,
+    ) -> result::Result<(), Error> {
+        let range_base = GuestAddress((pfn as u64) << VIRTIO_BALLOON_PFN_SHIFT);
+        let range_len = 1 << VIRTIO_BALLOON_PFN_SHIFT;
+
+        let page_size: u64 = get_page_size();
+        if page_size == 1 << VIRTIO_BALLOON_PFN_SHIFT {
+            return Self::release_memory_range(memory, range_base, range_len);
+        }
+
+        if pbp.is_none() {
+            *pbp = Some(PartiallyBalloonedPage::new());
+        }
+
+        let mut addr = range_base.0;
+        let mut len = range_len as i64;
+        while len > 0 {
+            if !pbp.as_ref().unwrap().pfn_match(addr) {
+                // We are trying to free memory region in a different pfn with current pbp. Flush addr
+                pbp.as_mut().unwrap().reset();
+                pbp.as_mut().unwrap().addr = align_page_size_up(addr);
```

Comment:
I think this should be `align_page_size_down()`

Reply:
Align down will affect the memory range that no need to be freed.

Reply:
Let's take a concrete example. Assume "page_size" is 2MB, and `addr` is `0x10_0000`. I believe the `pbp` should hold a bitmap to track address between `0x0` and `0x20_0000` (e.g. align down), not `0x20_0000` to `0x40_0000` (e.g. align up). This is also how you check `pfn_match(addr)`. Right?

Reply:
OK, make sense.